File: classic-puzzle-easy/csharp/logic-gates.cs
using System;
using System.Collections.Generic;

class Solution
{
    static void Main(string[] args)
    {
        int n = int.Parse(Console.ReadLine());
        int m = int.Parse(Console.ReadLine());

        Dictionary<string, string> inputSignals = new Dictionary<string, string>();
        List<Tuple<string, string, string, string>> outputSpecifications = new List<Tuple<string, string, string, string>>();
        List<string> outputResults = new List<string>();

        // Read input signals
        for (int i = 0; i < n; i++)
        {
            string[] inputs = Console.ReadLine().Split(' ');
            string inputName = inputs[0];
            string inputSignal = inputs[1];
            inputSignals[inputName] = inputSignal;
        }

        // Read output specifications
        for (int i = 0; i < m; i++)
        {
            string[] inputs = Console.ReadLine().Split(' ');
            string outputName = inputs[0];
            string type = inputs[1];
            string inputName1 = inputs[2];
            string inputName2 = inputs[3];

            outputSpecifications.Add(Tuple.Create(outputName, type, inputName1, inputName2));
        }

        // Compute output signals
        foreach (var spec in outputSpecifications)
        {
            string outputSignal = ComputeOutputSignal(inputSignals[spec.Item3], inputSignals[spec.Item4], spec.Item2);
            outputResults.Add($"{spec.Item1} {outputSignal}");
        }

        // Print output signals
        foreach (var result in outputResults)
        {
            Console.WriteLine(result);
        }
    }

    // Method to compute output signal
    static string ComputeOutputSignal(string inputSignal1, string inputSignal2, string gateType)
    {
        var outputSignal = new char[inputSignal1.Length];
        for (int i = 0; i < inputSignal1.Length; i++)
        {
            bool signal1 = inputSignal1[i] == '-';
            bool signal2 = inputSignal2[i] == '-';
            bool output;

            switch (gateType)
            {
                case "AND":
                    output = signal1 & signal2;
                    break;
                case "OR":
                    output = signal1 | signal2;
                    break;
                case "XOR":
                    output = signal1 ^ signal2;
                    break;
                case "NAND":
                    output = !(signal1 & signal2);
                    break;
                case "NOR":
                    output = !(signal1 | signal2);
                    break;
                case "NXOR":
                    output = !(signal1 ^ signal2);
                    break;
                default:
                    throw new ArgumentException("Invalid gate type.");
            }

            outputSignal[i] = output ? '-' : '_';
        }
        return new string(outputSignal);
    }
}

File: classic-puzzle-easy/rust/logic-gates.rs
use std::collections::HashMap;
use std::io;

fn main() {
    let mut input = String::new();

    // Read the number of input signals
    io::stdin().read_line(&mut input).unwrap();
    let n: usize = input.trim().parse().unwrap();
    input.clear();

    // Read the number of output signals
    io::stdin().read_line(&mut input).unwrap();
    let m: usize = input.trim().parse().unwrap();
    input.clear();

    let mut input_signals = HashMap::new();

    // Read the input signals
    for _ in 0..n {
        io::stdin().read_line(&mut input).unwrap();
        let parts: Vec<&str> = {
            input.trim().split_whitespace().collect()
        };
        let signal_vec = parts[1].chars().map(|c| c == '-').collect::<Vec<bool>>();
        input_signals.insert(parts[0].to_string(), signal_vec);
        input.clear();
    }

    // Read the output specifications and compute results
    for _ in 0..m {
        io::stdin().read_line(&mut input).unwrap();
        let parts: Vec<&str> = {
            input.trim().split_whitespace().collect()
        };

        let output_name = parts[0].to_string();
        let gate_type = parts[1];
        let input_signal1 = input_signals.get(parts[2]).unwrap();
        let input_signal2 = input_signals.get(parts[3]).unwrap();

        // Perform gate operation and create the output signal
        let output_signal: String = input_signal1
            .iter()
            .zip(input_signal2.iter())
            .map(|(&b1, &b2)| match gate_type { //closure ( | ) is an anonymous function that can capture variables from its surrounding scope
                "AND" => if b1 && b2 { '-' } else { '_' },
                "OR" => if b1 || b2 { '-' } else { '_' },
                "XOR" => if b1 ^ b2 { '-' } else { '_' },
                "NAND" => if !(b1 && b2) { '-' } else { '_' },
                "NOR" => if !(b1 || b2) { '-' } else { '_' },
                "NXOR" => if !(b1 ^ b2) { '-' } else { '_' },
                _ => unreachable!(),
            })
            .collect();

        println!("{} {}", output_name, output_signal);
        input.clear();
    }
}
